signing/signing.go

```diff
@@ -1,20 +1,363 @@
+	entry, err := cosign.GetTlogEntry(ctx, rekorClient, uuid)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error fetching tlog entry: %v", err)
+	}
+
+	// Verify tlog entry to make sure it is actually in the log.
+	err = cosign.VerifyTLogEntry(ctx, rekorClient, entry)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error verifying tlog entry: %v", err)
+	}
+
+	// Extract certificate.
+	certs, err := extractCerts(entry)
+	if err != nil || len(certs) == 0 {
+		return "", "", "", "", fmt.Errorf("error extracting certificate from entry: %v", err)
+	}
+	if len(certs) > 1 {
+		return "", "", "", "", fmt.Errorf("multiple certificates found for the entry: %v", err)
+	}
+
+	cert := certs[0]
+
+	// Verify that cert isn't expired.
+	if err = cosign.CheckExpiry(cert, time.Unix(*entry.IntegratedTime, 0)); err != nil {
+		return "", "", "", "", fmt.Errorf("cosign certificate is expired: %v", err)
+	}
+
+	// Get repo reference & path from cert.
+	for _, ext := range cert.Extensions {
+		// OID source: https://github.com/sigstore/fulcio/blob/96ef49cc7662912ba37d46f738757e8d8d5b5355/docs/oid-info.md#L33
+		// TODO: retrieve these by name.
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.6" {
+			repoRef = string(ext.Value)
+		}
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.5" {
+			repoPath = string(ext.Value)
+		}
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.3" {
+			repoSHA = string(ext.Value)
+		}
+	}
+
+	if len(repoRef) == 0 || len(repoPath) == 0 {
+		return "", "", "", "", fmt.Errorf("error extracting repo ref or path from certificate %v", err)
+	}
+
+	// Get workflow job ref from the certificate.
+	certURIs := cert.URIs
+	if len(certURIs) == 0 {
+		return "", "", "", "", errors.New("error: certificate has no URIs")
+	}
+	workflowRef := certURIs[0].Path
+	if len(workflowRef) == 0 {
+		return "", "", "", "", errors.New("error: workflow path is empty")
+	}
+
+	// Remove repo path from workflow filepath.
+	ind := strings.Index(workflowRef, repoPath) + len(repoPath) + 1
+	workflowPath = workflowRef[ind:]
+
+	// Remove repo ref tag from workflow filepath.
+	workflowPath = workflowPath[:strings.Index(workflowPath, "@")]
+
+	return repoPath, repoRef, repoSHA, workflowPath, nil
+}
+
+func verifyScorecardWorkflow(workflowContent string) error {
+	// Verify workflow contents using actionlint.
+	workflow, lintErrs := actionlint.Parse([]byte(workflowContent))
+	if lintErrs != nil {
+		return fmt.Errorf("actionlint errors parsing workflow: %v", lintErrs)
+	}
+
+	// Verify that there are no global env vars or defaults.
+	if workflow.Env != nil || workflow.Defaults != nil {
+		return errors.New("workflow contains global env vars or defaults")
+	}
+
+	// Verify that the only global permission set is read-all.
+	if workflow.Permissions.All.Value != "read-all" {
+		return errors.New("workflow permission isn't read-all")
+	}
+
+	// Extract main (analysis) job
+	jobsMap := workflow.Jobs
+	jobs := reflect.ValueOf(jobsMap).MapKeys()
+
+	if len(jobs) != 1 {
```

Comment:
there can be more than one job. Remember the link I gave u a while back with the implementation I had? Could you use that to make it more reliable?

Reply:
Fixed

Reply:
forgot to push?

Reply:
Working on this rn. How do we know which job is the right one? Can we assume that the id will be "analysis" or should we check if the name is "Scorecards analysis"?

Reply:
I think you need to walk all jobs and find the one that calls scorecard's action + verify the other jobs don't call it; and that the other jobs don't have the token-id permissions.

Reply:
I added logic to fetch the job that calls ossf/scorecard-action, and also make sure no other jobs have id-token set to write.